tempesta_fw/t/unit/sse_parser/sse_parser.c

```diff
@@ -0,0 +1,1552 @@
+    __m128i cmp1, cmp2, carry;
+    int     mask, n, maxn;
+
+    carry = _mm_setzero_si128();
+    maxn  = ts->iterations;
+    n     = 0;
+    //we must fit this cycle into 28 instructions to make if fast!
+    while(n < maxn) {
+        //prepare comparison masks for first bundle
+        cmp1 = _mm_shuffle_epi8(vec, ts->data[n++]);
+        cmp1 = _mm_cmpeq_epi32(cmp1, ts->data[n++]);
+        cmp2 = _mm_packs_epi32(carry, cmp1);
+        carry = cmp1;
+        cmp2 = _mm_and_si128(cmp2, cc.tokencmp);
+        cmp2 = _mm_hadd_epi16(cmp2, cmp2);
+        cmp2 = _mm_hadd_epi16(cmp2, cmp2);
+        cmp2 = _mm_hadd_epi16(cmp2, cmp2);
+        cmp2 = _mm_and_si128(cmp2, ts->data[n]);
+        cmp2 = _mm_cmpeq_epi16(cmp2, ts->data[n++]);
+        mask = 0xFF & _mm_movemask_epi8(cmp2);
+        cmp2 = _mm_and_si128(cmp2, ts->data[n++]);
+        if (mask) break;
+    }
+    if (mask) {
+        cmp2 = _mm_hadd_epi16(cmp2, cmp2);
+        cmp2 = _mm_hadd_epi16(cmp2, cmp2);
+        cmp2 = _mm_hadd_epi16(cmp2, cmp2);
+        mask = _mm_extract_epi16(cmp2, 0);
+    }
+    return mask;
+}
+
+#define MakeCode(len, cd) ((len<<16)|cd)
+
+int tokenSetLength(const char ** tokens)
+{
+    //count number of words required for map
+    int i, j, k;
+    for(i = 0, k = 0; tokens[i]; ++i) {
+        for(j = 0; tokens[i][j]; ++j);
+        k += (j+3)>>2;
+    }
+    //round number of words
+    return 16+64*((k+3)>>2);
+}
+
+TokenSet * initTokenSet(const char ** tokens,
+                 void               * buf,
+                 int                  bufsize)
+{
+    static const unsigned char cmap[4][4]= {
+        {0x10, 0x18, 0x1c, 0x1e},
+        {0x20, 0x30, 0x38, 0x3c},
+        {0x40, 0x60, 0x70, 0x78},
+        {0x80, 0xc0, 0xe0, 0xf0}
+    };
+    //check if buf is aligned
+    if (((long)buf) & 0xF) return NULL;
+
+    TokenSet * ts = (TokenSet*)buf;
+    //count number of words required for map
+    int i, j, k, p, nwords;
+    for(i = 0, k = 0; tokens[i]; ++i) {
+        for(j = 0; tokens[i][j]; ++j);
+        k += (j+3)>>2; 
+    }
+    //round number of words 
+    nwords = (k+3)>>2;
+    //check buffer size
+    if (bufsize < (16 + 64*nwords)) return NULL;
+
+    ts->iterations = nwords*4;
+    unsigned char * bytes = (unsigned char*)ts->data;
+    for(i = 0, k = 0; tokens[i]; ++i) {
+        for(j = 0; tokens[i][j]; ++j) {
+            bytes[k]    = j;
+            bytes[k+16] = tokens[i][j];
+            bytes[k+32] = 0xFF;
+            bytes[k+48] = 0;
+            p = k;
+            if ((k & 0xF) == 0xF) {
+                k = k + 49;
+            } else {
+                k = k + 1;
+            }
+        }
+
+        while(k & 3) {
+            bytes[k]    = 0xFF;
+            bytes[k+16] = 0;
+            bytes[k+32] = 0xFF;
+            bytes[k+48] = 0;
+
+            if ((k & 0xF) == 0xF) {
+                k = k + 49;
+            } else {
+                k = k + 1;
+            }
+        }
+
+        //WEIRD CODE!!!!
```

Comment:
senseless bad comment. Basically you should have been comment the loop calculations and `cmap` values.